Fix part of #13162: Add schema for handlers in story_editor and story_viewer

mahendra1290 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13162.
2. This PR does the following: Add handler class schema for `StoryEditorPage`, `StoryPublishHandler`, `StoryPageDataHandler` and `StoryProgressHandler`. And remove `StoryPage` from payload_validators. 

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma pre submit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![Screenshot from 2021-09-24 15-43-49](https://user-images.githubusercontent.com/42772952/134659569-2feef54d-bb8d-4538-bb7d-0c2f81d92b32.png)
![Screenshot from 2021-09-24 15-44-04](https://user-images.githubusercontent.com/42772952/134659581-655722ab-2c8d-4dc4-9966-d52dfae90712.png)


<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

mahendra1290 committed:
add handler args schema

mahendra1290 committed:
add handler args schema

mahendra1290 committed:
remove handler name

mahendra1290 committed:
refactor story_viewer handlers

vojtechjelinek commented on core/controllers/story_editor.py:
```diff
@@ -34,6 +34,17 @@
     """The editor page for a single story."""
 
+    URL_PATH_ARGS_SCHEMAS = {
+        'story_id': {
+            'schema': {
+                'type': 'basestring'
+            }
+        }
+    }
```
Should we impose more rules on `story_id`?

vojtechjelinek replied:
ditto elsewhere

mahendra1290 replied:
I think we can use this as a reference. https://github.com/mahendra1290/oppia/blob/293831f30952d28774d1bd3b6c35da7b3eef84f7/core/domain/story_domain.py#L786

![Screenshot from 2021-09-24 22-44-47](https://user-images.githubusercontent.com/42772952/134715052-76ecdd58-047b-4953-aa1f-605b80e31e94.png)

vojtechjelinek replied:
Agreed

vojtechjelinek commented on core/controllers/story_editor.py:
```diff
@@ -144,14 +155,31 @@ class StoryPublishHandler(base.BaseHandler):
-        new_story_status_is_public = self.payload.get(
+        new_story_status_is_public = self.normalized_payload.get(
             'new_story_status_is_public')
         if not isinstance(new_story_status_is_public, bool):
             raise self.InvalidInputException
```
This is no longer needed, since you introduced the validation in the schema.

vojtechjelinek replied:
ditto elsewhere

mahendra1290 replied:
done

mahendra1290 replied:
done

vojtechjelinek commented on core/controllers/story_editor.py:
```diff
@@ -144,14 +155,31 @@ class StoryPublishHandler(base.BaseHandler):
+                'schema': {
+                    'type': 'bool'
+                },
+                'default_value': None
```
Why is default value `None`?

mahendra1290 replied:
There was default_value at some other places so I added it here

vojtechjelinek replied:
Does it make sense here? What was the behaviour before if this thing wasn't set?

vojtechjelinek commented on core/controllers/story_viewer.py:
```diff
@@ -85,6 +128,26 @@ class StoryProgressHandler(base.BaseHandler):
+        'story_url_fragment': {
+            'schema': SCHEMA_FOR_STORY_URL_FRAGMENT
+        },
+        'node_id': {
+            'schema': {
+                'type': 'basestring'
+            }
+        }
```
Can more rules be introduce here?

mahendra1290 replied:
We can use this as a reference.
https://github.com/mahendra1290/oppia/blob/293831f30952d28774d1bd3b6c35da7b3eef84f7/core/domain/story_domain.py#L254
![Screenshot from 2021-09-24 22-53-45](https://user-images.githubusercontent.com/42772952/134715983-cd4a6fe7-e688-4911-9f60-ced895d096b7.png)

vojtechjelinek replied:
agreed

mahendra1290 replied:
when using regex for validation some tests fail. As we are passing 'invalid_node' as node_id which is not a valid node id so we never get a 404 error and our tests fail. 

![Screenshot from 2021-09-25 12-12-28](https://user-images.githubusercontent.com/42772952/134761679-796a17a2-ddbb-46ad-994f-ad71d9ee31bb.png)
added this validation

![Screenshot from 2021-09-25 12-08-31](https://user-images.githubusercontent.com/42772952/134761584-ea967586-2d43-4ae8-a97f-e445e2c0673e.png)

![Screenshot from 2021-09-25 12-09-56](https://user-images.githubusercontent.com/42772952/134761624-b4d85b3a-f5cc-4794-99e2-b99d2f99ef8c.png)

Nik-09 replied:
Yes you can modify there to raise 400 as we are expecting it to fail by Schema Validation System.

mahendra1290 replied:
done

vojtechjelinek reviewed (changes requested):
This looks great! Left a few comments.

mahendra1290 committed:
address comments: remove unnecessary code

mahendra1290 commented:
Hi, @vojtechjelinek I found these schemas in `constants.ts`.

We have separately defined MAX_CHAR_LIMIT and also regex for URL_FRAGMENT this may lead to issues in future.

![Screenshot from 2021-09-24 23-01-31](https://user-images.githubusercontent.com/42772952/134716985-95ac9268-3bc4-48fb-a511-53fc5bda8b3a.png)

mahendra1290 commented:
Also, how can we put this regex in `constants.ts`. Because we cannot import NODE_ID_PREFIX in `constants.ts`. only way is to  directly use` 'node_[0-9]+' `as a regex in constants.ts.

![Screenshot from 2021-09-24 22-53-45](https://user-images.githubusercontent.com/42772952/134717504-e574cb00-dc0d-4f6b-a820-90644351dd71.png)

vojtechjelinek commented:
> Hi, @vojtechjelinek I found these schemas in `constants.ts`.
> 
> We have separately defined MAX_CHAR_LIMIT and also regex for URL_FRAGMENT this may lead to issues in future.
> 
> ![Screenshot from 2021-09-24 23-01-31](https://user-images.githubusercontent.com/42772952/134716985-95ac9268-3bc4-48fb-a511-53fc5bda8b3a.png)

Why it might lead to problems?



> Also, how can we put this regex in `constants.ts`. Because we cannot import NODE_ID_PREFIX in `constants.ts`. only way is to directly use`'node_[0-9]+'`as a regex in constants.ts.
> 
> ![Screenshot from 2021-09-24 22-53-45](https://user-images.githubusercontent.com/42772952/134717504-e574cb00-dc0d-4f6b-a820-90644351dd71.png)

We generally only put stuff that is used both in backend and frontend to constants.ts, if it is used in multiple places in backend we put it into feconf.py, if it is used just in one file we put it into that file.

mahendra1290 commented:
It may lead to an issue because one may forget to update char limit at both places for URL fragments.

mahendra1290 committed:
add story_id validation

mahendra1290 commented:
I added validation for story_id in schema_utils.
![Screenshot from 2021-09-25 13-01-15](https://user-images.githubusercontent.com/42772952/134763125-5368cd60-9936-42ca-a52b-510089fb95c1.png)

But facing this error while pushing changes please help @vojtechjelinek 

![Screenshot from 2021-09-25 12-55-56](https://user-images.githubusercontent.com/42772952/134763143-e2f93359-b4e0-4287-add8-5062fb283971.png)
![Screenshot from 2021-09-25 13-03-06](https://user-images.githubusercontent.com/42772952/134763184-40b1abe5-a4d3-4e9d-abfe-8c5c70f8b8b9.png)

mahendra1290 commented:
Hi, @vojtechjelinek can you please help me out and guide me forward?

vojtechjelinek commented:
@mahendra1290 Sorry for the late reply, I have discussed this with @Nik-09 and he will look into this, I think the main problem is that you import `story_domain` in schema_utils.py

mahendra1290 commented:
> @mahendra1290 Sorry for the late reply, I have discussed this with @Nik-09 and he will look into this, I think the main problem is that you import `story_domain` in schema_utils.py

yes it causes problem

Nik-09 commented on core/controllers/story_viewer.py:
```diff
@@ -33,12 +33,55 @@
 import utils
 
 
+SCHEMA_FOR_CLASSROOM_URL_FRAGMENT = {
```
I think you can these schemas, you can introduce extra validator but try to use it from constants.ts.
https://github.com/oppia/oppia/blob/a0c63b07712a0cfb34a0cc5d4de8aaceeb709b9c/assets/constants.ts#L6350

mahendra1290 replied:
Sure I can use this.

mahendra1290 replied:
done

Nik-09 reviewed (commented):
@mahendra1290 Please take a look. Thank you.

Nik-09 commented:
@mahendra1290 Based on the image -- can you explain why ```story domain``` is imported in ```schema_utils```, just wanted to know what task you are expecting ?
Thank you.

mahendra1290 commented:
> @mahendra1290 Based on the image -- can you explain why ```story domain``` is imported in ```schema_utils```, just wanted to know what task you are expecting ?
> Thank you.

I was trying to reuse ```require_valid_story_id``` for validation from ```story domain``` ```Story``` model. As @vojtechjelinek mentioned to add more validation.

Nik-09 commented:
> > @mahendra1290 Based on the image -- can you explain why `story domain` is imported in `schema_utils`, just wanted to know what task you are expecting ?
> > Thank you.
> 
> I was trying to reuse `require_valid_story_id` for validation from `story domain` `Story` model. As @vojtechjelinek mentioned to add more validation.

@mahendra1290 Yes I understand, but why in schema utils ? If you want to use that class then pass directly into schema.
Can you confirm whether this helps or not, may be we can chat on mail thread.
Thank you.

mahendra1290 commented:
> > > @mahendra1290 Based on the image -- can you explain why `story domain` is imported in `schema_utils`, just wanted to know what task you are expecting ?
> > > Thank you.
> > 
> > 
> > I was trying to reuse `require_valid_story_id` for validation from `story domain` `Story` model. As @vojtechjelinek mentioned to add more validation.
> 
> @mahendra1290 Yes I understand, but why in schema utils ? If you want to use that class then pass directly into schema or if you only want to use `require_valid_story_id` then you first have to form an object from that class. You can copy the logic instead or try to impose any different rule in order to validate strory_id. Can you confirm whether this helps or not, may be we can chat on mail thread. Thank you.

We can use a direct method for validation in case of 'object_dict' type.
![Screenshot from 2021-10-01 22-06-49](https://user-images.githubusercontent.com/42772952/135656617-d2d951ff-f5cc-444f-ae5f-e2ff876fcf6f.png)

for other, we use the validation method defined in schema_utils. `require_valid_story_id` is a class method so we need to create object.
![Screenshot from 2021-10-01 22-03-57](https://user-images.githubusercontent.com/42772952/135656787-4c141331-95ad-4005-8fdb-dbba85d8042a.png)

I can shift story_id length to constants and then we can use max_len min_len validation from schema_utlis.

The main advantage of using require_valid_story_id was that in the future we may update that method and everything will work fine no need to update code at two places, I think I am thinking more.

mahendra1290 committed:
Remove is_valid_story_id

remove is_valid_story_id method and add has_length validation to
check for checking exact length.

mahendra1290 committed:
Use constant for validating `story id`

use STORY_ID_LENGTH at all places where 12 was used.

mahendra1290 committed:
Merge remote-tracking branch 'upstream/develop' into schema-for-handler-class-arguments

mahendra1290 committed:
use predefined constants

mahendra1290 committed:
add test for has_length method

mahendra1290 committed:
Add validation for node_id

add regex for node_id validation and update expected error code
from 404 to 400

mahendra1290 commented:
Hi @Nik-09 I have made some changes PTAL!

nithusha21 reviewed (approved):
LGTM for codeowner file

vojtechjelinek commented on core/controllers/story_viewer.py:
```diff
@@ -39,6 +40,14 @@ class StoryPageDataHandler(base.BaseHandler):
 
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    URL_PATH_ARGS_SCHEMAS = {
+        'classroom_url_fragment': constants.SCHEMA_FOR_CLASSROOM_URL_FRAGMENTS,
+        'topic_url_fragment': constants.SCHEMA_FOR_TOPIC_URL_FRAGMENTS,
+        'story_url_fragment': constants.SCHEMA_FOR_STORY_URL_FRAGMENTS,
```
Wait I see these params here but they are not used in the function below, do you know why is that?

mahendra1290 replied:
according to the wiki, we add all URL path elements into URL_PATH_ARGS_SCHEMAS whether these are used or not. you can check `PracticeSessionsPageDataHandler` for reference.

![Screenshot from 2021-10-05 15-01-40](https://user-images.githubusercontent.com/42772952/135997936-eaf238f4-815d-424b-9766-f3da431b30fb.png)

vojtechjelinek replied:
Ah, okay. Good point.

vojtechjelinek reviewed (changes requested):
Thanks! One more comment.

DubeySandeep commented on core/schema_utils.py:
```diff
@@ -416,6 +416,21 @@ def has_length_at_most(obj: List[str], max_value: int) -> bool:
+        `value` elements.
+
+        Args:
+            obj: list(str). A list of strings.
```
Are we sure that the list of elements in this object is always a string?

mahendra1290 replied:
I used signature of `has_length_at_least` and `has_length_at_most` for reference.

![Screenshot from 2021-10-05 14-16-28](https://user-images.githubusercontent.com/42772952/135991053-439be44b-36cb-48cf-a4ba-e62dfbe54ae8.png)

DubeySandeep reviewed (approved):
LGTM for the code owner file `core/schema_utils.py`, left one minor comment, PTAL!

mahendra1290 commented:
Hi @nithusha21 @Nik-09 @vojtechjelinek , there is one more issue with the addition of node_id validation. As when the node_id is not valid (format is not valid) we get 400 error but when node_id format is valid and it doesn't exist we get 404. We need to add more tests.

Update: I have added tests and now all checks pass PTAL!

mahendra1290 committed:
Merge remote-tracking branch 'upstream/develop' into schema-for-handler-class-arguments

mahendra1290 committed:
Fix tests related to node_id

add test for invalid node_schema and update node
doesn't exist test with expected error code 404.

vojtechjelinek reviewed (approved):
LGTM!

mahendra1290 commented:
Hi, @aks681 Please review this PR and let me know about any changes required. Thanks!

aks681 reviewed (approved):
Lgtm as codeowner
